File: services/orders-service/app/models/order.py
import enum
import uuid
from decimal import Decimal

from sqlalchemy import (
    DECIMAL,
    Boolean,
    CheckConstraint,
    DateTime,
    Enum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Priority(str, enum.Enum):
    baixa = "baixa"
    media = "media"
    alta = "alta"
    urgente = "urgente"


class OrderStatus(str, enum.Enum):
    pendente = "pendente"
    em_andamento = "em_andamento"
    concluido = "concluido"
    cancelado = "cancelado"


# Valid status transitions (state machine from data-model.md)
VALID_TRANSITIONS: dict[OrderStatus, set[OrderStatus]] = {
    OrderStatus.pendente: {OrderStatus.em_andamento, OrderStatus.cancelado},
    OrderStatus.em_andamento: {OrderStatus.concluido, OrderStatus.cancelado},
    OrderStatus.concluido: set(),   # final state
    OrderStatus.cancelado: set(),   # final state
}


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    customer_name: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_email: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    priority: Mapped[Priority] = mapped_column(
        Enum(Priority, name="priority_enum"), nullable=False, default=Priority.media
    )
    status: Mapped[OrderStatus] = mapped_column(
        Enum(OrderStatus, name="orderstatus_enum"),
        nullable=False,
        default=OrderStatus.pendente,
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    total_amount: Mapped[Decimal] = mapped_column(
        DECIMAL(12, 2), nullable=False, default=Decimal("0.00")
    )
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    created_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    items: Mapped[list["OrderItem"]] = relationship(
        "OrderItem", back_populates="order", cascade="all, delete-orphan", lazy="selectin"
    )

    __table_args__ = (
        Index("ix_orders_status", "status"),
        Index("ix_orders_priority", "priority"),
        Index("ix_orders_created_at", "created_at"),
        Index("ix_orders_created_by", "created_by"),
    )


class OrderItem(Base):
    __tablename__ = "order_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("orders.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False)
    unit_price: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)
    subtotal: Mapped[Decimal] = mapped_column(DECIMAL(12, 2), nullable=False)

    order: Mapped["Order"] = relationship("Order", back_populates="items")

    __table_args__ = (
        CheckConstraint("quantity >= 1", name="ck_order_items_quantity_positive"),
        CheckConstraint("unit_price >= 0", name="ck_order_items_unit_price_nonneg"),
        CheckConstraint("subtotal >= 0", name="ck_order_items_subtotal_nonneg"),
        Index("ix_order_items_order_id", "order_id"),
    )

File: frontend/orders-mfe/src/types/order.ts
export type OrderStatus = 'pendente' | 'em_andamento' | 'concluido' | 'cancelado'
export type Priority = 'baixa' | 'media' | 'alta' | 'urgente'

export interface OrderItem {
  name: string
  quantity: number
  unit_price: string
}

export interface Order {
  id: string
  customer_name: string
  customer_email: string
  description: string
  items: OrderItem[]
  total_amount: string
  status: OrderStatus
  priority: Priority
  notes: string | null
  created_by: string
  created_at: string
  updated_at: string
}

export interface OrderListResponse {
  items: Order[]
  total_count: number
  page: number
  page_size: number
}

export interface CreateOrderPayload {
  customer_name: string
  customer_email: string
  description: string
  items: { name: string; quantity: number; unit_price: string }[]
  priority: Priority
  notes?: string
}

export interface UpdateOrderPayload {
  status?: OrderStatus
  priority?: Priority
  notes?: string
}

export const VALID_TRANSITIONS: Record<OrderStatus, OrderStatus[]> = {
  pendente: ['em_andamento', 'cancelado'],
  em_andamento: ['concluido', 'cancelado'],
  concluido: [],
  cancelado: [],
}

export const STATUS_LABELS: Record<OrderStatus, string> = {
  pendente: 'Pendente',
  em_andamento: 'Em Andamento',
  concluido: 'Concluído',
  cancelado: 'Cancelado',
}

export const PRIORITY_LABELS: Record<Priority, string> = {
  baixa: 'Baixa',
  media: 'Média',
  alta: 'Alta',
  urgente: 'Urgente',
}
